crates/shared/src/price_estimation/sanitized.rs

```diff
@@ -0,0 +1,311 @@
+use super::{Estimate, PriceEstimating, PriceEstimationError, Query};
+use crate::bad_token::BadTokenDetecting;
+use crate::price_estimation::gas::GAS_PER_WETH_UNWRAP;
+use anyhow::Result;
+use model::order::BUY_ETH_ADDRESS;
+use primitive_types::H160;
+use std::collections::HashSet;
+use std::sync::Arc;
+
+/// Verifies that buy and sell tokens are supported and handles
+/// ETH as buy token appropriately.
+pub struct SanitizedPriceEstimator<T: PriceEstimating> {
+    inner: T,
+    bad_token_detector: Arc<dyn BadTokenDetecting>,
+    native_token: H160,
+}
+
+type EstimationResult = Result<Estimate, PriceEstimationError>;
+
+enum EstimationProgress {
+    TrivialSolution(EstimationResult),
+    AwaitingEthEstimation,
+    AwaitingErc20Estimation,
```

Comment:
ubernit: Instead of splitting into two vectors, could this be:
```suggestion
    AwaitingEthEstimation(EstimationResult),
    AwaitingErc20Estimation(EstimationResult),
```

That way the logic in `merge_trivial_and_forwarded_estimations` is slightly easier (it doesn't have to `.next().expect(...)`).

Reply:
I don't think this is possible with my current approach.
The goal is to use the bulk estimation API of the wrapped estimator. Because the `SanitizedPriceEstimator` is supposed to do as much trivial work as possible before passing queries it can't solve to the wrapped estimator.
Because the bulk estimating API takes a `&[Query]` instead of an iterator I have to have 2 `Vec` if I'm not mistaken. (1 Vec holds the trivial results, 1 Vec gets passed into `PriceEstimating::estimates(queries: &[Query])`)
The Vec holding the trivial results is supposed to have these place holders so that I'm able to easily merge the other estimates back in order.
So if I put the `EstimationResult` into the enum variants I still have to have the second Vec (otherwise I'm not able to call the bulk estimate API) while also have to create those `EstimationResult` values out of thin air to put them in the place holders.

Reply:
Ah right, of course.